### PR TITLE
Remove ChatFormat, InstructTemplate, old message converters

### LINKED ISSUE #1839: Remove deprecated `_converters.py` file

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Closes #1839 and #1849.

#### Changelog
What are the changes made in this PR?
* Delete ChatFormat, InstructTemplate, and all references
* Delete `torchtune/data/_converters.py`. These are replaced by the transforms in `_messages.py`
* Delete the old dataset tutorial. All topics have already been covered in Basics section of documentation, and the tutorial refers to outdated APIs. Only sample packing was still new, so I moved that to its own section under Basics.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: docs/source/basics/packing.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
+.. _packing_usage_label:
```
Comment (SalmanMohammadi) [inline docs/source/basics/packing.rst:1]: you snuck this in here you sneaky lil man

i love it

File: recipes/generate.py [hunk 2]
```diff
@@ -100,54 +99,23 @@ def _setup_model(
...
-
-        # dict.items() will respect order for Python >= 3.7
-        else:
-            messages = [Message(role=k, content=v) for k, v in prompt.items()]
-            messages += [Message(role="assistant", content="")]
-            if chat_format:
-                chat_format = _get_component_from_path(chat_format)
-                messages = chat_format.format(messages)
-            return self._tokenizer.tokenize_messages(messages)[0]
+        messages = [
```
Comment (SalmanMohammadi) [inline recipes/generate.py:108]: noobq: Is this identical to free-form generation when a prompt template isn't provided?

Comment (joecummings) [reply recipes/generate.py:108]: No, this assumes instruct-based finetuned models I believe.

File: docs/source/api_ref_data.rst [hunk 3]
```diff
@@ -37,18 +33,6 @@ Types
...
-.. autosummary::
-    :toctree: generated/
-    :nosignatures:
-
-    get_sharegpt_messages
-    get_openai_messages
-
 .. _message_transforms_ref:
 
 Message transforms
```
Comment (joecummings) [inline docs/source/api_ref_data.rst:38]: nit: Can we call these `ToMessage` transforms to convey that they convey immediately that they convert data to message format?

File: docs/source/basics/packing.rst (hunk 1 above)
Comment (joecummings) [reply docs/source/basics/packing.rst:1]: heh nice

File: docs/source/basics/packing.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+.. code-block:: bash
+
+    # Command line
+    tune run full_finetune_single_device --config llama3_2/1B_full_single_device \
+    dataset.packed=True tokenizer.max_seq_len=8192
+
+torchtune will automatically handle document masking and relative position IDs when sample packing is enabled
+to prevent different irrelevant samples from cross-attending. This is done via PyTorch's `Flex Attention <https://pytorch.org/blog/flexattention/#document-maskingjagged-sequences>`_,
+which enables the use of flash attention with non-causal masks. If your hardware does not support Flex Attention
+(for CUDA devices, it must be Turing or above), standard SDPA with ememory-efficient attention will be used as a fallback,
```
Comment (joecummings) [inline docs/source/basics/packing.rst:51]: typo: ememory-efficient -> memory-efficient

File: docs/source/basics/packing.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
+.. _packing_usage_label:
+
+==============
+Sample packing
+==============
+
+You can use sample packing with any of the single dataset builders by passing in
```
Comment (joecummings) [inline docs/source/basics/packing.rst:7]: Do we want to give 2-3 sentences on what sample packing is or no?

File: recipes/configs/generation.yaml [hunk 6]
```diff
@@ -27,11 +27,10 @@ tokenizer:
   _component_: torchtune.models.llama2.llama2_tokenizer
   path: /tmp/Llama-2-7b-hf/tokenizer.model
   max_seq_len: null
+  prompt_template: null
 
 # Generation arguments; defaults taken from gpt-fast
 prompt: "Tell me a joke?"
```
Comment (joecummings) [inline recipes/configs/generation.yaml:33]: I'm wondering if we should just adopt a format like generation_v2 here where we have the default be to make it clear:

```yaml
prompt:
	user: Tell me a joke.
```

**Review by joecummings — APPROVED** [review]
No big complaints - thx!

File: recipes/configs/generation.yaml (hunk 6 above)
Comment (RdoubleA) [reply recipes/configs/generation.yaml:33]: yeah that works. was just trying to make the minimal changes needed since we are migrating to generate_c2

File: docs/source/basics/packing.rst (hunk 1 above)
Comment (RdoubleA) [reply docs/source/basics/packing.rst:1]: ok I just now realized the joke

File: docs/source/api_ref_data.rst (hunk 3 above)
Comment (RdoubleA) [reply docs/source/api_ref_data.rst:38]: I've used "Message transforms" throughout the docs, so I'll leave updating all those references for a future PR and keep this as is
